text/0000-freeze-uninit.md

```diff
@@ -0,0 +1,217 @@
+Freeze allows these limited Rust Programs to convert uninitialized data into useless-but-initialized bytes.
+
+Examples of uses:
+1. The major use for freeze is to read padding bytes of structs. This can be used for a [generic wrapper around standard atomic types](https://docs.rs/atomic/latest/atomic/struct.Atomic.html). 
+2. SIMD Code using masks can load a large value by freezing the bytes, doing lanewise arithmetic operations, then doing a masked store of the initialized elements. With additional primitives not specified here, this can allow for efficient partial load operations which prevent logical operations from going out of bounds (such a primitive could be defined to yield uninit for the lane, which could then be frozen).
+3. Low level libraries, such as software floating-point implementations, used to provide operations for compilers where uninit is considered a valid value for the provided operations.
+    * Along the same lines, a possible fast floating-point operation set that yields uninit on invalid (such as NaN or Infinite) results, stored as `MaybeUninit`, then frozen upon return as `f32`/`f64`.
+    * Note that such operations require compiler support, and these operations are *not* defined by this RFC.
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Two new library interfaces are defined for unsafe code users:
+
+```rust
+// in module `core::ptr`
+pub unsafe fn read_freeze<T>(ptr: *const T) -> T;
+
+impl<T> *const T{
+    pub unsafe fn read_freeze(self) -> T;
+}
+impl<T> *mut T{
+    pub unsafe fn read_freeze(self) -> T;
+}
+
+// in module `core::mem`
+impl<T> MaybeUninit<T>{
+    pub fn freeze(self) -> Self;
+}
+```
+
+`read_freeze` is an operation that loads a number of bytes from (potentially uninitialized) memory, replacing every byte in the source that is uninitialized with an arbitrary-but-initialized value.  
+The function has similar safety constraints to `core::ptr::read`, in that the source pointer must be *dereferenceable* and well aligned, as well as pointing to a valid value of type `T` (after replacing uninitialized bytes).  
+The same-name functions on the raw pointer types are identical to the free function version, and are provided for convience. 
+Only the bytes of the return value are frozen. The bytes behind the pointer argument are unmodified.  
+**The `read_freeze` operation does not freeze any padding bytes of `T` (if any are present), and those are set to uninit after the read as usual.**
+
+`MaybeUninit::freeze` is a safe, by-value version of `read_freeze`. It takes in a `MaybeUninit` and yields an initialized value, which is either exactly `self` if it is already initialized, or some arbitrary value if it is not. 
```

Comment:
It may be worth noting that the term "initialized" here refers to the notion of an Abstract Byte in memory being initialized. This is not to be confused with the notion of `MaybeUninit` being "initialized" in the sense that `assume_init()` is safe to call. The text says that `MaybeUninit::<bool>::uninit().freeze()`  is initialized but it is obviously not safe to call `assume_init` on this.